pico_w/mqtt/mqtt_client/mqtt_client.c

```diff
@@ -0,0 +1,161 @@
+//
+// Created by elliot on 25/05/24.
+//
+#include "pico/stdlib.h"
+#include "hardware/gpio.h"
+#include "hardware/irq.h"
+#include "hardware/adc.h"
+#include "pico/cyw43_arch.h"
+#include "lwip/apps/mqtt.h"
+
+
+// Temperature
+#define TEMPERATURE_UNITS 'C'
+
+// WiFi
+#define WIFI_SSID "YOUR_WIFI_SSID"
+#define WIFI_PASSWORD "YOUR_WIFI_PASSWD"
+// MQTT
+#define PORT  1883
+#define MQTT_CLIENT_ID "clientID"
+#define MQTT_BROKER_IP "YOUR_BROKER_IP"
+
+
+
+typedef struct {
+    mqtt_client_t* mqtt_client_inst;
+    struct mqtt_connect_client_info_t mqtt_client_info;
+    uint8_t data[MQTT_OUTPUT_RINGBUF_SIZE];
+    uint8_t topic[100];
+    uint32_t len;
+} MQTT_CLIENT_DATA_T;
+
+MQTT_CLIENT_DATA_T *mqtt;
+
+
+
+/* References for this implementation:
+ * raspberry-pi-pico-c-sdk.pdf, Section '4.1.1. hardware_adc'
+ * pico-examples/adc/adc_console/adc_console.c */
+float read_onboard_temperature(const char unit) {
+
+    /* 12-bit conversion, assume max value == ADC_VREF == 3.3 V */
+    const float conversionFactor = 3.3f / (1 << 12);
+
+    float adc = (float)adc_read() * conversionFactor;
+    float tempC = 27.0f - (adc - 0.706f) / 0.001721f;
+
+    if (unit == 'C') {
+        return tempC;
+    } else if (unit == 'F') {
+        return tempC * 9 / 5 + 32;
+    }
+
+    return -1.0f;
+}
+
+void control_led(bool on) {
+    // Public state on /state topic and on/off led board
```

Comment:
There's quite a few code-comments where you've said "Public blah blah topic" - were these _supposed_ to say "Publish blah blah topic" ?